streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -0,0 +1,523 @@
+            final int oldCumValueSize = cumulativeValueSizes.get(index);
+
+            final boolean needsMinTsUpdate = isLastIndex(index);
+            unpackedReversedTimestampAndValueSizes.set(index, new TimestampAndValueSize(timestamp, value.valueSize()));
+            cumulativeValueSizes.set(index, oldCumValueSize - oldValueSize + value.value().length);
+            truncateDeserHelpersToIndex(index);
+
+            // update serialization and other props
+            final int segmentTimestampIndex = 2 * TIMESTAMP_SIZE + index * (TIMESTAMP_SIZE + VALUE_SIZE);
+            segmentValue = ByteBuffer.allocate(segmentValue.length - oldValueSize + value.value().length)
+                .put(segmentValue, 0, segmentTimestampIndex)
+                .putLong(timestamp)
+                .putInt(value.valueSize())
+                .put(segmentValue, segmentTimestampIndex + TIMESTAMP_SIZE + VALUE_SIZE, segmentValue.length - (segmentTimestampIndex + TIMESTAMP_SIZE + VALUE_SIZE) - oldCumValueSize)
+                .put(value.value())
+                .put(segmentValue, segmentValue.length - oldCumValueSize + oldValueSize, oldCumValueSize - oldValueSize)
+                .array();
+
+            if (needsMinTsUpdate) {
+                minTimestamp = timestamp;
+                ByteBuffer.wrap(segmentValue, TIMESTAMP_SIZE, TIMESTAMP_SIZE).putLong(TIMESTAMP_SIZE, minTimestamp);
+            }
+        }
+
+        @Override
+        public byte[] serialize() {
+            return segmentValue;
+        }
+
+        private void initializeWithRecord(final ValueAndValueSize value, final long validFrom, final long validTo) {
+            this.nextTimestamp = validTo;
+            this.minTimestamp = validFrom;
+            this.segmentValue = ByteBuffer.allocate(TIMESTAMP_SIZE * 3 + VALUE_SIZE + value.value().length)
+                .putLong(nextTimestamp)
+                .putLong(minTimestamp)
+                .putLong(validFrom)
+                .putInt(value.valueSize())
+                .put(value.value())
+                .array();
+            this.isEmpty = false;
+            resetDeserHelpers();
+        }
+
+        private void resetDeserHelpers() {
+            deserIndex = -1;
+            unpackedReversedTimestampAndValueSizes = new ArrayList<>();
+            cumulativeValueSizes = new ArrayList<>();
+        }
+
+        private void truncateDeserHelpersToIndex(final int index) {
+            deserIndex = index;
+            unpackedReversedTimestampAndValueSizes.subList(index + 1, unpackedReversedTimestampAndValueSizes.size()).clear();
+            cumulativeValueSizes.subList(index + 1, cumulativeValueSizes.size()).clear();
+        }
+
+        private boolean isLastIndex(final int index) {
+            if (deserIndex == -1
+                || unpackedReversedTimestampAndValueSizes.get(deserIndex).timestamp != minTimestamp) {
```

Comment:
Hm yeah I see your confusion. This method relies on the fact that `index` has already been deserialized, which means it'd be equally valid to replace this whole method implementation with 
```
return unpackedReversedTimestampAndValueSizes.get(index).timestamp == minTimestamp;
```
The way that it's currently written instead says: 
```
if (segment is not fully deserialized) {
    return false; // we don't even know what the last index is in this case
}
return index == lastIndex;
```
This is "safer" since it relies less on the fact that `index` has already been deserialized, but still implicitly uses that fact because the method does not try to deserialize up to `index` if that's not already the case.

In light of that, let me replace this method implementation with the simplified version, and leave a comment to clear this up.